client/client.js

```diff
@@ -78,67 +79,85 @@ new Vue({
       })
     },
     initialiseStatshistory: function(){
-      this.statshistory = [];
-      for (var i = 0; i < numXValues; i++) {
-      	this.statshistory.push(
-      	  {
-      	    Queue: 0,
-      	    Running: 0,
-      	    Complete: 0,
-      	    FunctionStatsMap: {}
-      	  }      	  
-      	)
-      } 
+      if (this.statshistory==null){
+        this.statshistory = [];
+        for (var i = 0; i < numXValues; i++) {
+          this.statshistory.push(
+            {
+              Queue: 0,
+              Running: 0,
+              Complete: 0,
+              FunctionStatsMap: {}
+            }      	  
+          )
+        } 
+      }
     },    
     loadStats: function(){
       var t = this;
-      $.ajax({
-        url: '/api/stats',
-        dataType: 'json',
-        success: (statistics) => {
-          t.stats = statistics;
-          if (t.statshistory==null){
-            t.statshistory = [statistics];
-          } else {
-            t.statshistory.push(statistics);
-            if (t.statshistory.length > numXValues){
-              t.statshistory.shift();
-            }
-          }        
-          
-          // do the stats contain a new function?
-          var previousKnownFunctions = Object.keys(t.statshistory[t.statshistory.length-2].FunctionStatsMap);
-          var nowKnownFunctions = Object.keys(t.statshistory[t.statshistory.length-1].FunctionStatsMap);
-          for (var j = 0; j < nowKnownFunctions.length; j++){
-            if (previousKnownFunctions.indexOf(nowKnownFunctions[j])==-1){
-              var newFunction = nowKnownFunctions[j];
-              // we have a new function: backfill all the earlier stats with zero values for this function
-              for (var k = 0; k < t.statshistory.length-1; k++){
-                t.statshistory[k].FunctionStatsMap[newFunction]={Queue:0, Running:0, Complete:0, Failed:0};
+      if (this.autorefresh) {
+        $.ajax({
+          url: '/api/stats',
+          dataType: 'json',
+          success: (statistics) => {
+            t.stats = statistics;
+            if (t.statshistory==null){
+              t.statshistory = [statistics];
+            } else {
+              t.statshistory.push(statistics);
+              if (t.statshistory.length > numXValues){
+                t.statshistory.shift();
+              }
+            }        
+            
+            // do the stats contain a new function?
```

Comment:
As we're using Babel you could look at using a Map (new in es6) here rather than a plain object. You could then use Map.has for checking membership and Map.forEach for checking the keys and backfilling. This would improve readability as the reader no longer has to parse the conditions of the `for`s and the usage of `indexOf`.

Reply:
I don't completely follow: which object are you suggesting I replace with a Map?

However I would rather postpone refactoring this section as it will be changed again in a few days when I change the format of the JSON data.